File: 30-Days-Of-Code/Day 12/inheritance.py
# given

class Person:
    def __init__(self, firstName, lastName, idNumber):
        self.firstName = firstName
        self.lastName = lastName
        self.idNumber = idNumber

    def printPerson(self):
        print("Name:{},{}".format(self.lastName, self.firstName))
        print("ID:{}".format(self.idNumber))

# start
class Student(Person):
    def __init__(self, firstName, lastName, idNumber, scores):
        self.firstName = firstName
        self.lastName = lastName
        self.idNumber = idNumber
        self.scores = scores

    def calculate(self):
        average = sum(self.scores) / len(self.scores)
        if average >= 90:
            grade = "O"
        elif average >= 80:
            grade = "E"
        elif average >= 70:
            grade = "A"
        elif average >= 55:
            grade = "P"
        elif average >= 40:
            grade = "D"
        else:
            grade = "T"
        return grade

# Given
line = input().split()
firstName = line[0]
lastName = line[1]
idNum = line[2]
numScores = int(input())
scores = list(map(int, input().split()))
s = Student(firstName, lastName, idNum, scores)
s.printPerson()
print("Grade:", s.calculate())

File: 30-Days-Of-Code/Day 12/inheritance.cpp
#include <iostream>
#include <vector>
#include <numeric>

using namespace std;

// Given
class Person {
    protected:
        string firstName;
        string lastName;
        int id;
    public:
        Person(string firstName, string lastName, int identification) {
            this->firstName = firstName;
            this->lastName = lastName;
            this->id = identification;
        }

        void printPerson() {
            cout << "Name: " << lastName << ", " << firstName << endl;
            cout << "ID: " << id << endl;
        }
};

// Start
class Student : public Person {
    private:
        vector<int> testScores;
    public:
        Student(string firstName, string lastName, int identification, vector<int> scores) :
            Person(firstName, lastName, identification) {
            this->testScores = scores;
        }

        char calculate() {
            float average =
                accumulate(testScores.begin(), testScores.end(), 0.0) / testScores.size();
            if (average >= 90) {
                return 'O';
            } else if (average >= 80) {
                return 'E';
            } else if (average >= 70) {
                return 'A';
            } else if (average >= 55) {
                return 'P';
            } else if (average >= 40) {
                return 'D';
            } else {
                return 'T';
            }
        }
};

// Given
int main() {
    string firstName;
    string lastName;
    int id;
    int numScores;
    cin >> firstName >> lastName >> id >> numScores;
    vector<int> scores;
    for (int i = 0; i < numScores; i++) {
        int tmpScore;
        cin >> tmpScore;
        scores.push_back(tmpScore);
    }
    Student *s = new Student(firstName, lastName, id, scores);
    s->printPerson();
    cout << "Grade: " << s->calculate() << endl;
    return 0;
}
